server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -506,6 +531,190 @@ private List<UploadedIndexMetadata> writeIndexMetadataParallel(ClusterState clus
         return result;
     }
 
+    private void uploadIndexPathAsync(
+        List<IndexMetadata> toUploadIndexPath,
+        CountDownLatch latch,
+        boolean isTranslogSegmentRepoSame,
+        List<Exception> exceptionList
+    ) throws IOException {
+        for (IndexMetadata indexMetadata : toUploadIndexPath) {
+            writeIndexPathAsync(indexMetadata, latch, isTranslogSegmentRepoSame, exceptionList);
+        }
+    }
+
+    private void writeIndexPathAsync(
```

Comment:
you need to refactor this code outside of RemoteClusterStateService class and plug it in here via interface. This brings in coupling with segment/ translog which has nothing to do with state as well.

Reply:
There are certain constraints like this code needs to run in parallel & in async same as upload index metadata code. The latch needs to be adjusted based on certain criteria. It was difficult to attain this without doing considerable change. Let me attempt this one more time.

Reply:
Can this code be plugged in a ClusterStateApplier? `IndicesClusterStateService.applyClusterState` method applies the changes in the cluster state to the cluster. From here there is a call to `createIndices` where you can get the indices which are getting created.

Reply:
@soosinha This particular code is no-op on ClusterManager node and is mostly relevant for Data nodes. I am going ahead with refactoring in the same class by introducing an interface and abstracting out the non-cluster manager code from the RemoteClusterStaetService class.